testing/unit/munit-mmio-c.cpp

```diff
@@ -0,0 +1,141 @@
+
+extern "C" {
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#include "intel-fpga.h"
+#include "gtest/gtest.h"
+#include "test_system.h"
+#include <opae/access.h>
+#include <opae/mmio.h>
+#include <sys/mman.h>
+#include "types_int.h"
+#include <cstdarg>
+#include <linux/ioctl.h>
+
+#undef FPGA_MSG
+#define FPGA_MSG(fmt, ...) \
+	printf("MOCK " fmt "\n", ## __VA_ARGS__)
+
+using namespace opae::testing;
+
+int mmio_ioctl(mock_object * m, int request, va_list argp){
+    int retval = -1;
+    errno = EINVAL;
+    struct fpga_port_region_info *rinfo = va_arg(argp, struct fpga_port_region_info *);
+    if (!rinfo) {
+      FPGA_MSG("rinfo is NULL");
+      goto out_EINVAL;
+    }
+    if (rinfo->argsz != sizeof(*rinfo)) {
+      FPGA_MSG("wrong structure size");
+      goto out_EINVAL;
+    }
+    if (rinfo->index != 0) {
+      FPGA_MSG("unsupported MMIO index");
+      goto out_EINVAL;
+    }
+    if (rinfo->padding != 0) {
+      FPGA_MSG("unsupported padding");
+      goto out_EINVAL;
+    }
+    rinfo->flags = FPGA_REGION_READ | FPGA_REGION_WRITE | FPGA_REGION_MMAP;
+    rinfo->size = 0x40000;
+    rinfo->offset = 0;
+    retval = 0;
+    errno = 0;
+out:
+    return retval;
+
+out_EINVAL:
+    retval = -1;
+    errno = EINVAL;
+    goto out;
+}
+
+class mmio_c_p
+    : public ::testing::TestWithParam<std::string> {
+ protected:
+  mmio_c_p() : tmpsysfs_("mocksys-XXXXXX"), handle_(nullptr) {}
+
+  virtual void SetUp() override {
+    ASSERT_TRUE(test_platform::exists(GetParam()));
+    platform_ = test_platform::get(GetParam());
+    system_ = test_system::instance();
+    system_->initialize();
+    tmpsysfs_ = system_->prepare_syfs(platform_);
+
+    ASSERT_EQ(fpgaGetProperties(nullptr, &filter_), FPGA_OK);
+    ASSERT_EQ(fpgaPropertiesSetObjectType(filter_, FPGA_ACCELERATOR), FPGA_OK);
+    ASSERT_EQ(fpgaEnumerate(&filter_, 1, tokens_.data(), tokens_.size(),
+                            &num_matches_),
+              FPGA_OK);
+    ASSERT_EQ(fpgaOpen(tokens_[0], &handle_, 0), FPGA_OK);
+    system_->register_ioctl_handler(FPGA_PORT_GET_REGION_INFO, mmio_ioctl);
+  }
+
+  virtual void TearDown() override {
+    EXPECT_EQ(fpgaDestroyProperties(&filter_), FPGA_OK);
+    if (handle_ != nullptr) EXPECT_EQ(fpgaClose(handle_), FPGA_OK);
+    if (!tmpsysfs_.empty() && tmpsysfs_.size() > 1) {
+      std::string cmd = "rm -rf " + tmpsysfs_;
+      std::system(cmd.c_str());
+    }
+    system_->finalize();
+  }
+
+  std::string tmpsysfs_;
+  fpga_properties filter_;
+  std::array<fpga_token, 2> tokens_;
+  fpga_handle handle_;
+  uint32_t num_matches_;
+  test_platform platform_;
+  test_system *system_;
+};
+
+TEST_P (mmio_c_p, test_map_mmio) {
+  uint64_t * mmio_ptr = NULL;
+  EXPECT_TRUE(((struct _fpga_handle*)handle_)->mmio_root == NULL);
+
+  ASSERT_EQ(FPGA_OK, fpgaMapMMIO(handle_, 0, &mmio_ptr));
+  EXPECT_FALSE(mmio_ptr == NULL);
+  EXPECT_FALSE(((struct _fpga_handle *)handle_)->mmio_root == NULL);
+  EXPECT_EQ(FPGA_OK, fpgaUnmapMMIO(handle_, 0));
```

Comment:
These can also be written as
```C++
EXPECT_EQ(lhs, nullptr);
// OR
EXPECT_NE(lhs, nullptr);
```
Which is comparing the left-hand side (lhs) against null.